lib/CL/devices/tbb/tbb.cc

```diff
@@ -0,0 +1,392 @@
+/* OpenCL device using the Intel TBB library (derived from the pthread device).
+
+   Copyright (c) 2011-2020 pocl developers
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to
+   deal in the Software without restriction, including without limitation the
+   rights to use, copy, modify, merge, publish, distribute, sublicense, and/or
+   sell copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS
+   IN THE SOFTWARE.
+*/
+
+#define _GNU_SOURCE
+#define __USE_GNU
+#include <sched.h>
+
+#include <algorithm>
+#include <assert.h>
+#include <pthread.h>
+#include <string.h>
+#include <stdlib.h>
+#include <errno.h>
+
+#ifndef _MSC_VER
+#  include <unistd.h>
+#else
+#  include "vccompat.hpp"
+#endif
+
+extern "C" {
+  #include "cpuinfo.h"
+  #include "topology/pocl_topology.h"
+}
+
+#include <tbb/task_arena.h>
+
+#include "common.h"
+#include "config.h"
+#include "devices.h"
+#include "pocl_mem_management.h"
+#include "pocl_runtime_config.h"
+#include "pocl_util.h"
+#include "tbb.h"
+#include "tbb_scheduler.h"
+#include "tbb_utils.h"
+#include "utlist.h"
+
+#ifndef HAVE_LIBDL
+#error tbb driver requires DL library
+#endif
+
+#ifdef OCS_AVAILABLE
+#include "pocl_llvm.h"
+#endif
+
+/**
+ * Per event data.
+ */
+struct event_data {
+  pthread_cond_t event_cond;
+};
+
+struct data {
+  /* Currently loaded kernel. */
+  cl_kernel current_kernel;
+  volatile uint64_t total_cmd_exec_time;
+};
+
+void
+pocl_tbb_init_device_ops(struct pocl_device_ops *ops)
```

Comment:
Can you do the "C-style inheritance" here and just assign pthread/basic functions to those that are identical or can be "overriden" by first calling it and then modifying after?

Reply:
See 3c81f27c58d9abfc58af9be860413dc8a0fd6ff1 and subsequent commits.